KRAB–zinc finger protein expression patterns during induced pluripotent stem cell to myeloid differentiation


Maria Luisa Orlandi (s1093374)
Supervisor: Dr. Colin Logie, department of Molecular Biology 
Daily Supervisor: Sybren Rinzema
03-02-2025 / 08-07-2025


Abstract
Krüppel-associated box domain–zinc finger (KZnF) proteins constitute the largest family of transcription factors in mammals. A subset of has been shown to be involved in epigenetic silencing of transposable elements (TE) during early embryonic development. Canonically, this is achieved by recruitment of the transcription co-repressor TRIM28 (KAP1) to promote constitutive H3K9me3 heterochromatin formation in embryonic stem cells when parental genome methylation is reset and DNA methylation-based gene expression repression is poorly operational (Habibi 2013).
This study aimed at profiling KZnF gene expression during the differentiation of induced pluripotent stem cells (iPSCs) into myeloid cells and investigating the regulatory context of KZnF. Given their involvement in TE repression during early embryogenesis, KZnF gene expression was hypothesised to be highest at the earliest stages and decrease along the differentiation. The analysis revealed two main expression patterns: 94 KZnF genes were biased towards embryonic stem cells and 37 towards later differentiation stages. 
The observed KZnF expression dynamics were found to occur before the induced hematopoietic stem cell stage (iHSC), since very few differences were observed between iHSC and the subsequent induced myeloid (iMye) stage. Analysis of TE targets further supported the idea that TE-mediated regulation is most active in early stages and repressed during intermediate stages. 
Gene Ontology term analysis suggested that early-stage KZnF genes are associated mainly with repressive function, reflecting their complex roles in regulating TE dynamics. Epigenetic profiling of selected KZnF genes revealed active promoter marks and H3K9me3 enrichment over zinc finger domains, potentially serving as protection against recombination rather than transcriptional repression. 
The limitations introduced by combining data from bulk RNA sequencing and pseudobulk data derived from single cells were also addressed, highlighting the need for a different approach in pseudobulk computation. 
Future studies should incorporate finer temporal resolution, hESC-based models, and single-cell multiomic approaches to better understand KZnF–TE interactions and their roles in human development.


The code shared in this repository generates the analyses discussed in the associated report. They can be ran independently from each other and are mostly corresponding to the different sections. 
